File: zig/src/day03.zig
const std = @import("std");
const log = std.log;
const testing = std.testing;

pub fn part1() usize {
    const content = @embedFile("data/day03.txt");
    const content_lines_len = 300;

    var item_list = std.BoundedArray(u8, content_lines_len).init(0) catch unreachable;

    var lines = std.mem.tokenizeScalar(u8, content, '\n');

    while (lines.next()) |line| {
        const midpoint = @divExact(line.len, 2);
        const first_compartment = line[0..midpoint];
        const second_compartment = line[midpoint..];

        const common_item_in_compartments = first_compartment[std.mem.indexOfAny(u8, first_compartment, second_compartment).?];
        item_list.appendAssumeCapacity(common_item_in_compartments);
    }

    const priority_sum = itemsToPriority(item_list.constSlice());
    log.info(
        "The sum of the priorities of common items is {[priority_sum]}",
        .{ .priority_sum = priority_sum },
    );
    return priority_sum;
}

fn itemsToPriority(item_list: []const u8) usize {
    const map_lower_ascii = 96; //@as(u8, '`');
    const map_upper_ascii = 64; //@as(u8, '@');
    const alphabet_count = @as(u8, 26);

    var priority_sum: usize = 0;

    for (item_list) |value| {
        if (std.ascii.isLower(value)) {
            const value_priority = value - map_lower_ascii;
            priority_sum += value_priority;
        } else {
            const value_priority = value - map_upper_ascii + alphabet_count;
            priority_sum += value_priority;
        }
    }
    return priority_sum;
}

test part1 {
    try testing.expectEqual(@as(usize, 8139), part1());
}

pub fn part2() usize {
    const content = @embedFile("data/day03.txt");
    const content_lines_len = 100;

    var badge_list = std.BoundedArray(u8, content_lines_len).init(0) catch unreachable;

    var elves_groups = std.mem.tokenizeScalar(u8, content, '\n');
    var line_count: usize = 0;

    while (elves_groups.next()) |elf| : (line_count += 1) {
        const first_elf = elf;
        const second_elf = elves_groups.next().?;
        const third_elf = elves_groups.next().?;

        for (first_elf) |item| {
            if (std.mem.indexOfScalar(u8, second_elf, item)) |common_item_index|
                if (std.mem.indexOfScalar(u8, third_elf, second_elf[common_item_index])) |badge_item_index| {
                    badge_list.appendAssumeCapacity(third_elf[badge_item_index]);
                    //Beacuse the items in the elves compartments aren't unique you have to break after puting the first
                    //copy into `badge_list` else you would have duplicates badges which leds to wrong output
                    break;
                } else continue
            else
                continue;
        }
    }

    const priority_sum = itemsToPriority(badge_list.constSlice());
    log.info(
        "The sum of the priorities of badges in the various 3 elf group is {[priority_sum]}",
        .{ .priority_sum = priority_sum },
    );
    return priority_sum;
}
test part2 {
    try testing.expectEqual(@as(usize, 2668), part2());
}

File: rust/src/day03.rs
use std::fs::read_to_string;

pub fn part1() -> Result<u32, Box<dyn std::error::Error>> {
    let content = read_to_string("src/data/day03.txt")?;
    let lines = content.lines();

    let common_item_in_various_rucksack_compartment =
        lines.into_iter().map(|line| {
            let midpoint = line.len().midpoint(0);
            let (first_compartment, second_compartment) =
                line.split_at(midpoint);
            let matching_item = first_compartment
                .chars()
                .filter(|&item| second_compartment.contains(item))
                .take(1)
                .next();

            matching_item.unwrap()
        });

    let priority_sum =
        items_to_priority(common_item_in_various_rucksack_compartment);
    println!("The priority sum of items in the rucksuck is {priority_sum}");

    Ok(priority_sum)
}

fn items_to_priority<T>(common_item_in_various_rucksack_compartment: T) -> u32
where
    T: std::iter::Iterator<Item = char>,
{
    let item_priority =
        common_item_in_various_rucksack_compartment.map(|item| {
            let lowercase_map = '`' as u32;
            let uppercase_map = '@' as u32;
            let lowercase_end = 26;
            if item.is_lowercase() {
                //convert lowercase char in item to a range of 1=a..z=26
                item as u32 - lowercase_map //'`' in ascii is 96
            } else {
                //convert uppercase char in item to a range of 27=A..Z=52
                (item as u32 - uppercase_map) + lowercase_end //'@' in ascii is 64
            }
        });

    item_priority.sum::<u32>()
}

#[test]
fn test_part1() {
    assert_eq!(8139, part1().unwrap());
}

pub fn part2() -> u32 {
    let content = read_to_string("src/data/day03.txt").unwrap();
    let lines = content.lines();

    // while let Ok(elve_group) = lines.next_chunk::<3>() {
    //     let [first, second, thrid] = elve_group;
    //     println!("{first}, {second},{thrid}");
    //     println!("\n");
    // }

    let group_badges =
        lines.array_chunks::<3>().map(|[first, second, thrid]| {
            let mut common_badge_item =
                first.chars().filter(|&items| second.contains(items));
            let next_common_badge_item = second
                .chars()
                .filter(|&items| thrid.contains(items))
                .collect::<Vec<char>>();

            let badge = common_badge_item
                .find(|badge| next_common_badge_item.contains(badge));
            badge.unwrap()
        });

    let priority_sum = items_to_priority(group_badges);

    println!("The sum of the priorities of the badge items is {priority_sum}");

    priority_sum
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_part1() {
        assert_eq!(8139, part1().unwrap());
    }

    #[test]
    fn test_part2() {
        assert_eq!(2668, part2());
    }
}
